server/src/test/java/com/tencent/rss/server/ShuffleUploaderTest.java

```diff
@@ -293,6 +296,45 @@ public void selectShuffleFiles() {
           .build();
       shuffleFileInfos = shuffleUploader.selectShuffleFiles(4, false);
       assertResult2(dataFiles, indexFiles, shuffleFileInfos);
+
+      when(mockDiskItem.getSortedShuffleKeys(false, 4))
+          .thenReturn(Lists.newArrayList(shuffleKey1, "zeroPartitionShuffleKey", "zeroSizeShuffleKey"));
+      when(mockDiskItem.getSortedShuffleKeys(false, 1))
+          .thenReturn(Lists.newArrayList(shuffleKey1));
+      shuffleUploader = new ShuffleUploader.Builder()
+          .diskItem(mockDiskItem)
+          .uploadThreadNum(2)
+          .uploadIntervalMS(3)
+          .uploadCombineThresholdMB(300)
+          .referenceUploadSpeedMBS(1)
+          .remoteStorageType(StorageType.HDFS)
+          .hdfsBathPath("hdfs://base")
+          .serverId("127.0.0.1-8080")
+          .maxShuffleSize(12)
+          .hadoopConf(new Configuration())
+          .build();
+      shuffleFileInfos = shuffleUploader.selectShuffleFiles(4, true);
+      assertEquals(2, shuffleFileInfos.size());
+      assertResult2(dataFiles, indexFiles, shuffleFileInfos);
+
+      when(mockDiskItem.getHighWaterMarkOfWrite()).thenReturn(95.0);
+      when(mockDiskItem.getLowWaterMarkOfWrite()).thenReturn(0.0);
+      when(mockDiskItem.getCapacity()).thenReturn(11L);
+      shuffleUploader = new ShuffleUploader.Builder()
+          .diskItem(mockDiskItem)
+          .uploadThreadNum(1)
+          .uploadIntervalMS(3)
+          .uploadCombineThresholdMB(300)
+          .referenceUploadSpeedMBS(1)
+          .remoteStorageType(StorageType.HDFS)
+          .hdfsBathPath("hdfs://base")
+          .serverId("127.0.0.1-8080")
+          .maxShuffleSize(9)
+          .hadoopConf(new Configuration())
+          .build();
+      shuffleFileInfos = shuffleUploader.selectShuffleFiles(4, true);
+      assertEquals(1, shuffleFileInfos.size());
+      assertResult4(dataFiles, indexFiles, shuffleFileInfos);
```

Comment:
we should a test case that the shuffle is very huge.

Reply:
This case is actually what you want, huge shuffle don't exceed the max segment num(aka thread num). And the last case is too.

Reply:
I mean a huge shuffle that have a huge partition.

Reply:
Added